src/modules/account/RadixTransferAccountSystem.ts

```diff
@@ -109,57 +130,116 @@ export default class RadixTransferAccountSystem implements RadixAccountSystem {
-
-            if (!(tokenId in this.balance) || !this.balance[tokenId]) {
-                this.balance[tokenId] = new BN(0)
+        for (const tokenId in currentBalance) {
+            if (!(tokenId in state.balance) || !state.balance[tokenId]) {
+                state.balance[tokenId] = new BN(0)
             }
 
-            this.balance[tokenId].iadd(transaction.balance[tokenId])
+            state.balance[tokenId].iadd(currentBalance[tokenId])
 
             // Token units
-            transaction.tokenUnitsBalance[tokenId] = RadixTokenDefinition.fromSubunitsToDecimal(transaction.balance[tokenId])
+            currentTokenUnitsBalance[tokenId] = RadixTokenDefinition.fromSubunitsToDecimal(currentBalance[tokenId])
 
-            if (!(tokenId in this.tokenUnitsBalance) || !this.balance[tokenId]) {
-                this.tokenUnitsBalance[tokenId] = new Decimal(0)
+            if (!(tokenId in state.tokenUnitsBalance) || !state.balance[tokenId]) {
+                state.tokenUnitsBalance[tokenId] = new Decimal(0)
             }
 
-            this.tokenUnitsBalance[tokenId] = this.tokenUnitsBalance[tokenId].add(transaction.tokenUnitsBalance[tokenId])
+            state.tokenUnitsBalance[tokenId] = state.tokenUnitsBalance[tokenId].add(currentTokenUnitsBalance[tokenId])
         }
 
+        return {
+            currentBalance,
+            currentParticipants,
+            currentTokenUnitsBalance
+        }
+    }
+
+    private processStoreAtom(atomUpdate: RadixAtomObservation) {
+        const atom = atomUpdate.atom
+
+
+        // Skip existing atoms
+        if (this.transactions.has(atom.getAidString())) {  // just to skip if we already processed this atom
+            return
+        }
+
+        const transactionUpdate: RadixTransactionUpdate = {
+            action: 'STORE',
+            aid: atom.getAidString(),
+            transaction: {
+                aid: atom.getAidString(),
+                balance: {},
+                tokenUnitsBalance: {},
+                fee: 0,
+                participants: {},
+                timestamp: atom.getTimestamp(),
+                message: '',
+                unique: atom.getParticlesOfType(RadixUniqueParticle).map(p => p.getRRI().toString()),
+            },
+        }
+
+        const transaction = transactionUpdate.transaction
+
+        // Get transaction message
+        if (atomUpdate.processedData.decryptedData
+            && atomUpdate.processedData.decryptedData.decryptionState !== RadixDecryptionState.CANNOT_DECRYPT) {
+            transaction.message = atomUpdate.processedData.decryptedData.data
+        }
+
+        //const consumables = atom.getSpunParticlesOfType(RadixTransferrableTokensParticle)
+        const state: TransferState = {
+            spentConsumables: this.spentConsumables,
+            unspentConsumables: this.unspentConsumables,
+            balance: this.balance,
+            tokenUnitsBalance: this.tokenUnitsBalance
+        }
```

Comment:
Do you not just end up modifying the actual instance state? This isn't a deep clone as far as I can tell

Reply:
In the transfer system, it currently mutates the state, yes. I will change that when I implement the pending state for it.